plugins/default/src/Ide/Plugin/Eval/GHC.hs

```diff
@@ -0,0 +1,198 @@
+>>> addPackagesTest pkgs = run (\_ -> (packageFlags <$>) <$> addPackages pkgs)
+
+>>> addPackagesTest []
+Right []
+
+>>> addPackagesTest ["base","array"]
+Right [-package base{package base True ([])},-package array{package array True ([])}]
+
+>>> addPackagesTest ["QuickCheck"]
+Left "<command line>: cannot satisfy -package QuickCheck\n    (use -v for more information)"
+
+>>> addPackagesTest ["notThere"]
+Left "<command line>: cannot satisfy -package notThere\n    (use -v for more information)"
+-}
+addPackages :: [String] -> Ghc (Either String DynFlags)
+addPackages pkgNames = gStrictTry $ modifyFlags (\df -> df{packageFlags = map expose pkgNames ++ packageFlags df})
+  where
+    expose name = ExposePackage ("-package " ++ name) (PackageArg name) (ModRenaming True []) -- -package-id filepath-1.4.2.1
+
+modifyFlags :: GhcMonad m => (DynFlags -> DynFlags) -> m DynFlags
+modifyFlags f = do
+    df <- getSessionDynFlags
+    _ <- setSessionDynFlags (f df)
+    getSessionDynFlags
+
+-- modifyFlags f = do
+--         modifyDynFlags f
+--         getSessionDynFlags
+
+{- | Add import to evaluation context
+
+>>> run $ \_ -> addImport "import Data.Maybe"
+Could not find module ‘Data.Maybe’
+Use -v (or `:set -v` in ghci) to see a list of the files searched for.
+
+>>> run $ \df -> addPackages ["base"] >> addImport "import Data.Maybe"
+[import Data.Maybe]
+
+>>> run $ \df -> addPackages ["base"] >> addImport "import qualified Data.Maybe as M"
+[import qualified Data.Maybe as M]
+-}
+addImport :: GhcMonad m => String -> m [InteractiveImport]
+addImport i = do
+    ctx <- getContext
+    -- dbgO "CONTEXT" ctx
+    idecl <- parseImportDecl i
+    setContext $ IIDecl idecl : ctx
+    -- ctx' <- getContext
+    -- dbg "CONTEXT'" ctx'
+    getContext
+
+{- | Add extension to interactive evaluation session
+>>> import GHC.LanguageExtensions.Type(Extension(..))
+>>> run $ \_ -> addExtension DeriveGeneric
+()
+-}
+addExtension :: GhcMonad m => Extension -> m ()
+addExtension ext =
+    modifySession $ \hsc -> hsc{hsc_IC = setExtension (hsc_IC hsc) ext}
+
+setExtension :: InteractiveContext -> Extension -> InteractiveContext
+setExtension ic ext = ic{ic_dflags = xopt_set (ic_dflags ic) ext}
+
+deriving instance Read Extension
+
+-- Partial display of DynFlags contents, for testing purposes
+instance Show DynFlags where
+    show df =
+        showSDocUnsafe . vcat . map (\(n, d) -> text (n ++ ": ") <+> d) $
+            [ ("extensions", ppr . extensions $ df)
+            , ("extensionFlags", ppr . EnumSet.toList . extensionFlags $ df)
+            , ("importPaths", vList $ importPaths df)
+            -- , ("includePaths", text . show $ includePaths df)
+            -- , ("packageEnv", ppr $ packageEnv df)
+            -- , ("packageFlags", vcat . map ppr $ packageFlags df)
+            -- ,("pkgDatabase",(map) (ppr . installedPackageId) . pkgDatabase $ df)
+            -- ,("pkgDatabase",text . show <$> pkgDatabase $ df)
+            ]
```

Comment:
Is it possible to turn this Orphan instance into a function `showDynFlags`?

Reply:
done